### DISCUSSION TITLE
Allow private methods on the server.

### DISCUSSION THREAD

**Discussion by Torgen** [meteor/meteor#13272] [Feature Requests]
Meteor 3 uses Node 20, and private fields and methods were supported on much earlier versions of Node: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties

However, if I try to use one, or use a node module that uses one, I get a "Class private methods are not enabled. Please add `@babel/plugin-transform-private-methods` to your configuration." message. (Actually trying to add that plugin to the babel section of package.json doesn't seem to help.)

**Comment by StorytellerCZ** [comment]
That is strange. If Node supports it, then it should work out of the box. The only thing I think of is if maybe it is hidden behind some flag or something similar. What does the Node.js documentation/release notes say?

**Reply by Torgen** [reply]
It never got as far as whether node could support it because Babel complained about it.

**Reply by make-github-pseudonymous-again** [reply]
I confirm that the compilation just halts there. Here is an example:

```console
=> Build failed:                              
   
   While building for web.browser:
   node_modules/pdfjs-dist/build/pdf.mjs: ./node_modules/pdfjs-dist/build/pdf.mjs: Class private methods are not enabled. Please add `@babel/plugin-transform-private-methods` to your
   configuration.
     682 |     return [xLow, yLow, xHigh, yHigh];
    683 |   }
   > 684 |   static #getExtremumOnCurve(x0, x1, x2, x3, y0, y1, y2, y3, t, minMax) {
        |   ^
    685 |     if (t <= 0 || t >= 1) {
    686 |       return;
    687 |     }
   
   While building for web.browser.legacy:
   node_modules/pdfjs-dist/build/pdf.mjs: ./node_modules/pdfjs-dist/build/pdf.mjs: Class private methods are not enabled. Please add `@babel/plugin-transform-private-methods` to your
   configuration.
     682 |     return [xLow, yLow, xHigh, yHigh];
    683 |   }
   > 684 |   static #getExtremumOnCurve(x0, x1, x2, x3, y0, y1, y2, y3, t, minMax) {
        |   ^
    685 |     if (t <= 0 || t >= 1) {
    686 |       return;
    687 |     }
   
   While building for os.linux.x86_64:
   node_modules/pdfjs-dist/build/pdf.mjs: ./node_modules/pdfjs-dist/build/pdf.mjs: Class private methods are not enabled. Please add `@babel/plugin-transform-private-methods` to your
   configuration.
     682 |     return [xLow, yLow, xHigh, yHigh];
    683 |   }
   > 684 |   static #getExtremumOnCurve(x0, x1, x2, x3, y0, y1, y2, y3, t, minMax) {
        |   ^
    685 |     if (t <= 0 || t >= 1) {
    686 |       return;
    687 |     }

```

**Reply by Torgen** [reply]
The funny thing is, it works in my own code; it's only in imported modules that it's a problem.

**Reply by Torgen** [reply]
In `babel-presets-meteor` the "modern" version of node is 8. The built-in version of Node for Meteor 3 is at least 20, but I suppose new meteor is more likely to work on old Node than old Meteor is to work on new node, since async always worked but Fiber doesn't anymore, That said, could a modern meteor app actually work on Node 8 anymore, or should that be bumped? Or perhaps it should vary based on the Meteor release of the host app?

**Reply by StorytellerCZ** [reply]
According to the comments in the file it seems to refer to async functions, we can try to bump that and see what it does.

**Reply by StorytellerCZ** [reply]
That said, here is the PR that we can test against: #13802

**Reply by Torgen** [reply]
I also noticed that while the readme for the `babel-compiler` package says you can add plugins in the .babelrc file, I tried it and it didn't work; it seems that babelrc is set to false in https://github.com/meteor/meteor/blob/018371ef65facdc0dc8eaf22e71af417ce483e1e/npm-packages/meteor-babel/options.js#L126

**Reply by nachocodoner** [reply]
Please, refer to this comment to check Babel status in Meteor, https://github.com/meteor/meteor/pull/13802#issuecomment-2996187240

[In latest Meteor 3.3 version](https://forums.meteor.com/t/faster-builds-in-meteor-3-3-modern-build-stack-with-swc-and-bundler-optimizations/63697/1), [by activating SWC](https://docs.meteor.com/about/modern-build-stack/transpiler-swc.html) / "modern": true, you can see your code supporting latest syntax like private methods as requested.

Apparently you will need to [add a project .swcrc config](https://docs.meteor.com/about/modern-build-stack/transpiler-swc.html#custom-swcrc) with:

```
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
      "privateMethod": true
    }
  }
}
```

**Reply by nachocodoner** [reply]
I can confirm that with SWC it worked for me. Not even activating privateMethod. Could you try?

![image](https://github.com/user-attachments/assets/ea5729e4-2749-46c7-a14e-8af2407f017b)

**Reply by Torgen** [reply]
Turning on SWC, it seems like transpilation succeeded but in --full-app mode the app crashed on startup with the error
```
I20250623-18:44:17.606(-7)? /tmp/meteor-test-runmpn0sm.t3yq9/.meteor/local/build/programs/server/packages/modules.js:1776
I20250623-18:44:17.606(-7)? const __filename = fileURLToPath(import.meta.url);
I20250623-18:44:17.606(-7)?                                         ^^^^
I20250623-18:44:17.606(-7)?
I20250623-18:44:17.606(-7)? SyntaxError: Cannot use 'import.meta' outside a module
I20250623-18:44:17.606(-7)?     at new Script (node:vm:117:7)
I20250623-18:44:17.606(-7)?     at createScript (node:vm:269:10)
I20250623-18:44:17.606(-7)?     at Object.runInThisContext (node:vm:317:10)
I20250623-18:44:17.606(-7)?     at /tools/static-assets/server/boot.js:414:32
I20250623-18:44:17.607(-7)?     at /tools/static-assets/server/boot.js:502:11
I20250623-18:44:17.607(-7)?     at Function.run (/tmp/meteor-test-runmpn0sm.t3yq9/.meteor/local/build/programs/server/tools/tool-env/profile.ts:651:14)
I20250623-18:44:17.607(-7)?     at startServerProcess (/tools/static-assets/server/boot.js:501:17)
I20250623-18:44:17.607(-7)?     at Object.<anonymous> (/tools/static-assets/server/boot.js:506:2)
I20250623-18:44:17.607(-7)?     at Object.<anonymous> (/tools/static-assets/server/boot.js:510:4)
I20250623-18:44:17.607(-7)?     at Module.<anonymous> (node:internal/modules/cjs/loader:1730:14)
```

**Reply by nachocodoner** [reply]
Do you mean when tunning the tests for your app? I'll need a small repository with your config to review the problem here. Going to try that mode in a small app though. But it could be likely something specific on your config.

Edit: By a chance do you use nodeModules.recompile as part of your app config. Again a small reproduction repo will help a lot 🙏

**Reply by Torgen** [reply]
I don't have a minimal reproduction since I'm trying to do this with my real app. https://github.com/Torgen/codex-blackboard/tree/cannot-use-import-meta is a branch with the change in it; I run the full app test with `TEST_BROWSER_DRIVER=playwright meteor npm run test.full`.

**Reply by nachocodoner** [reply]
Thanks, I’ll check out that repo and see if I can debug the issue.

Could you clarify: do those tests pass when "modern": false? Or is this related to using private methods or packages that neither modern nor legacy Meteor build stacks support?

Are your tests stable on any branch of your app?

**Reply by Torgen** [reply]
The tests work at master (non-modern transpiler, hubot version 9).
If you update hubot to version 11 or later and leave the transpiler as non-modern, the build fails with the private method issue.
With hubot at version 11 or higher and modern transpiler, the build succeeds, but the app crashes on startup and never gets as far as the tests.
I haven't tried modern transpiler without updating hubot.

**Reply by Torgen** [reply]
Hubot 9 with modern transpiler passes. Probably https://github.com/hubotio/hubot/blame/c9aff4b31d4a5a83252458e3e40416c31dda6cfb/src/Robot.mjs#L19 is where SWC is generating bad output. I don't know if Babel would do better if it was able to get past the private methods.

**Reply by nachocodoner** [reply]
> The tests work at master (non-modern transpiler, hubot version 9).

Thanks for these clarifications. I see the issue is the Hubot upgrade to v11, which uses MJS and ESM modules. Meteor doesn’t fully support these, not even via SWC, it doesn’t handle ESM module intrinsics. This mirrors the issue with [Upgrading pdfjs-dist to v4](https://forums.meteor.com/t/upgrading-pdfjs-dist-to-v4/61599).

We’ll add full ESM compilation support in the [Meteor-RSPack integration](https://forums.meteor.com/t/meteor-rspack-integration-a-modern-bundler-meets-meteor-3-4/63696), using a modern tool to ensure compatibility with libraries that use ESM.

**Update:** In the meantime, [joely suggest a way to do a pre-compilation to ensure the file is ready for Meteor](https://forums.meteor.com/t/upgrading-pdfjs-dist-to-v4/61599/31)

**Reply by Torgen** [reply]
I'm still having the issue with `import.meta` in mjs files after updating to Meteor 3.4.1.
